tracer/src/Datadog.Trace/HttpOverStreams/ChunkedEncoding/ChunkedEncodingWriteStream.cs

```diff
@@ -0,0 +1,117 @@
+
+    private readonly Stream _innerStream = innerStream;
+    private readonly byte[] _chunkSizeBuffer = new byte[8]; // max length of Int32 as UTF-8 hex
+
+    public override async Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+    {
+        // don't want to ever send a zero-length terminator unless we're at the end
+        if (count == 0)
+        {
+            // Don't write if nothing was given, especially since we don't want to accidentally send a 0 chunk,
+            // which would indicate end of body.  Instead, just ensure no content is stuck in the buffer.
+            await _innerStream.FlushAsync(cancellationToken).ConfigureAwait(false);
+            return;
+        }
+
+        // write the chunked encoding header
+        var bytesWritten = WriteChunkedEncodingHeaderToBuffer(_chunkSizeBuffer, count);
+        await _innerStream.WriteAsync(_chunkSizeBuffer, 0, bytesWritten, cancellationToken).ConfigureAwait(false);
+
+        // add the new line
+        await _innerStream.WriteAsync(CrLfBytes, offset: 0, count: 2, cancellationToken).ConfigureAwait(false);
+
+        // add the content
+        await _innerStream.WriteAsync(buffer, offset, count, cancellationToken).ConfigureAwait(false);
+
+        // add the extra new line
+        await _innerStream.WriteAsync(CrLfBytes, offset: 0, count: 2, cancellationToken).ConfigureAwait(false);
+    }
+
+    public override void Write(byte[] buffer, int offset, int count)
+    {
+        // don't want to ever send a zero-length terminator unless we're at the end
+        if (count == 0)
+        {
+            // Don't write if nothing was given, especially since we don't want to accidentally send a 0 chunk,
+            // which would indicate end of body.  Instead, just ensure no content is stuck in the buffer.
+            _innerStream.Flush();
+            return;
+        }
+
+        // write the chunked encoding header
+        var bytesWritten = WriteChunkedEncodingHeaderToBuffer(_chunkSizeBuffer, count);
+        _innerStream.Write(_chunkSizeBuffer, 0, bytesWritten);
+
+        // add the new line
+        _innerStream.WriteAsync(CrLfBytes, offset: 0, count: 2);
+
+        // flush the content
+        _innerStream.Write(buffer, offset, count);
+
+        // add the extra new line
+        _innerStream.Write(CrLfBytes, offset: 0, count: 2);
+    }
+
+    public Task FinishAsync()
+    {
+        // Send 0 byte chunk to indicate end, then final CrLf
+        return _innerStream.WriteAsync(FinalChunkBytes, 0, 5);
+    }
+
+    public override void Close()
+    {
+        // Override so that we don't dispose the inner stream
+    }
+
+    protected override void Dispose(bool disposing)
+    {
+        // Override so that we don't dispose the inner stream
+    }
+
+#if NETCOREAPP
+    public override ValueTask DisposeAsync()
+    {
+        // Override so that we don't dispose the inner stream
+        return default;
```

Comment:
Maybe less confusing to return `ValueTask.CompletedTask`?

Reply:
If only `ValueTask.CompletedTask` existed in .NET Core 3.1... 😶